### PR TITLE
Prepare lf to use stpv (image previews)

### PR DESCRIPTION
**Pull request by Naheel-Azawy**
Finally image previews working! #134

This does not add the functionality of previewing images to LF, it only provide the right environment for the scripts to do their work.

From my quick test, it works. It needs [THIS](https://github.com/Naheel-Azawy/stpv), `set previewer stpv`, and LF launched as @gokcehan mentioned in the issue comments:

```
fid="$(mktemp)"
lf -command '$printf $id > '"$fid"'; stpvimg --listen $id &' -last-dir-path="$tmp" "$@"
stpvimg --end "$(cat $fid)"
rm $fid
```

There are two `TODO`s though that I'm not so sure about yet. @gokcehan, if you can check and let me know.

### CODE REVIEW

**Comment by gokcehan** [comment]
@Naheel-Azawy Isn't `syscall` a too low level library for this purpose? Is there any other way to do this without that package? I don't want to break things for windows and other platforms.

**Comment by Naheel-Azawy** [comment]
@gokcehan I'm not so sure of how to do that in Windows... I don't want to say there is no other way to get the exit status in go as I'm not a pro in it, but from my search that was the only way I found.

Maybe find another way other than the exit status? Maybe set some special keyword as output that will prevent caching?

**Comment by gokcehan** [comment]
@Naheel-Azawy Sorry for late reply.

If I understand correctly, they have added [ProcessState.ExitCode](https://golang.org/pkg/os/#ProcessState.ExitCode) in Go 1.12. I haven't tried it but I think we can make use of this.

The other thing is that we should avoid an extra call for people who don't use this feature as you noted in your comments. I was thinking of a separate option for this like `cleaner` which we can check before calling `nav.previewClear` so we can avoid the extra goroutine. Also current way assumes a flag `--clear` which can reduce the usability for tools which does not have this flag.

**Comment by Naheel-Azawy** [comment]
@gokcehan `ProcessState.ExitCode` did the job! I've updated it. 

> Also current way assumes a flag --clear which can reduce the usability for tools which does not have this flag.

Any suggestions?..

**Comment by gokcehan** [comment]
@Naheel-Azawy I meant to say that if we use a separate option like `cleaner`, then we don't have to hard code the `--clean` flag in the code. It should still be possible to create a wrapper script as such:

    #!/bin/sh
    mpv --clean $id $@

Or we could simply implement `cleaner` option to accept flags as such:

    set cleaner "mpv:--clean:$id"

(`:` is currently used to separate multiple values in our options though I was thinking of changing it to `,` instead. Also I think we may need to do something about `$id` to make it work as above)

**Comment by Naheel-Azawy** [comment]
@gokcehan Sorry for my slow response but I'm too busy as usual...
Anyway, done with the cleaner option! For now, I used the wrapper option...

I don't think separating shell commands with `:` or `,` is a really good idea. It makes sense for options but I don't see it continent for commands. I think some kind of a mini parser should be implemented. I thought maybe we can do `exec.Command("sh", "-c", whatever)`, but this is kind of dirty and will probably scare Windows. So the only option I can think of is implementing a little parser. Any suggestions?

One last thing, the only TODO left, for [`newNav`](https://github.com/gokcehan/lf/pull/208/files#diff-a66cc59eb6bfe62b7b3b69cd52b84747R37), am I doing the right thing? I think not. Maybe I should pass a reference to the window better? Sometimes the previews don't show up exactly in the right place. I didn't dig into this problem but you're the master her so maybe you can figure out what's happening easily?

**Comment by gokcehan** [comment]
@Naheel-Azawy Cleaner option seems to be fine. We can add it to the documentation later on after some experimentation. Wrapping approach should be fine for now. We can think about adding a mini parser later on. I think a mini parser could also be used to combine `shell` and `shellopts` options.

Passing `width`, `x`, and `y` seems to be right, thought it is a little bit confusing to store these in `nav`. The reason is that `height` is shared among all windows but the others are not. Instead of storing these in `nav` I think we can simply access them just before passing these arguments to the script. It's a cleanup thing so I can do this after merge. Also do you think we should switch the places of arguments `x` and `y` since `height` comes before `width`?

I still haven't tried this fork as I don't have those image previewers installed yet. I can take a look at the position issue you mention when I try this.

**Comment by Naheel-Azawy** [comment]
@gokcehan 
> We can think about adding a mini parser later on. I think a mini parser could also be used to combine `shell` and `shellopts` options.

Totally agree. But I unfortunately don't think I can implement this right now as I have tons of other things to do...

> Also do you think we should switch the places of arguments `x` and `y` since `height` comes before `width`?

Well, the convenient order in my point of view is `x, y, width, height`. You know... `x` always come before `y`... But I kept `height` at the beginning not to ruin any existing scripts. So I think it's fine for any arrangement you like.

Anyway, overall I think I'm happy where this is going! I hope to see this on master as soon as possible!

**Comment by gokcehan** [comment]
@Naheel-Azawy I have finally tried this patch. So `height` is updated throughout the code when something changes but `x`, `y`, and `width` are not. This is probably why images are sometimes not shown in the right place. As I said before, maybe it is better to access them just before we pass them to the script to get the most updated values. These values are not used in nav calculations anyways.

The bigger issue is that this patch does not handle caching properly. For example if I `set period 1`, then images flicker at each second. Same goes for operations that should be noop. For example if I'm at the top of the files and use `up` command, image will still need to be refreshed. These issues are discussed in the issue with more detail.

Lastly, sometimes I get a crash. I think there might be a race somewhere. I have got the following stack trace from a crash:

    goroutine 41 [select]:
    main.(*ui).pollEvent(0xc0001a4240, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
            /tmp/asd/lf/ui.go:789 +0x123
    main.(*ui).readExpr.func1(0xc00018e180, 0xc0001a4240)
            /tmp/asd/lf/ui.go:942 +0xd0
    created by main.(*ui).readExpr
            /tmp/asd/lf/ui.go:938 +0x75

    goroutine 42 [IO wait]:
    internal/poll.runtime_pollWait(0x7f6495996d98, 0x72, 0xffffffffffffffff)
            /usr/lib/go/src/runtime/netpoll.go:182 +0x56
    internal/poll.(*pollDesc).wait(0xc0002a4098, 0x72, 0x1000, 0x1000, 0xffffffffffffffff)
            /usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x9b
    internal/poll.(*pollDesc).waitRead(...)
            /usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
    internal/poll.(*FD).Read(0xc0002a4080, 0xc000368000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
            /usr/lib/go/src/internal/poll/fd_unix.go:169 +0x19b
    net.(*netFD).Read(0xc0002a4080, 0xc000368000, 0x1000, 0x1000, 0xc00020bcd8, 0x44354c, 0x1000)
            /usr/lib/go/src/net/fd_unix.go:202 +0x4f
    net.(*conn).Read(0xc0002a8000, 0xc000368000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
            /usr/lib/go/src/net/net.go:177 +0x69
    bufio.(*Scanner).Scan(0xc00020bf48, 0x625e00)
            /usr/lib/go/src/bufio/scan.go:213 +0xa4
    main.readExpr.func1(0xc00018e1e0)
            /tmp/asd/lf/client.go:66 +0x3f6
    created by main.readExpr
            /tmp/asd/lf/client.go:50 +0x58

    goroutine 49 [select]:
    github.com/nsf/termbox-go.Init.func1()
            /home/gokce/go/pkg/mod/github.com/nsf/termbox-go@v0.0.0-20190325093121-288510b9734e/api.go:95 +0x3
    32
    created by github.com/nsf/termbox-go.Init
            /home/gokce/go/pkg/mod/github.com/nsf/termbox-go@v0.0.0-20190325093121-288510b9734e/api.go:92 +0x7
    39

    goroutine 448 [IO wait]:
    internal/poll.runtime_pollWait(0x7f64940a2dd8, 0x72, 0xffffffffffffffff)
            /usr/lib/go/src/runtime/netpoll.go:182 +0x56
    internal/poll.(*pollDesc).wait(0xc00006e798, 0x72, 0x1001, 0x1000, 0xffffffffffffffff)
            /usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x9b
    internal/poll.(*pollDesc).waitRead(...)
            /usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
    internal/poll.(*FD).Read(0xc00006e780, 0xc000369000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
            /usr/lib/go/src/internal/poll/fd_unix.go:169 +0x19b
    os.(*File).read(...)
            /usr/lib/go/src/os/file_unix.go:263
    os.(*File).Read(0xc0000a6018, 0xc000369000, 0x1000, 0x1000, 0x40d4b9, 0xc0001e8050, 0x50)
            /usr/lib/go/src/os/file.go:108 +0x70
    bufio.(*Scanner).Scan(0xc000059f48, 0xc0001e8050)
            /usr/lib/go/src/bufio/scan.go:213 +0xa4
    main.(*nav).preview(0xc0001a4360)
            /tmp/asd/lf/nav.go:426 +0x99b
    created by main.(*nav).loadReg
            /tmp/asd/lf/nav.go:448 +0x9f

**Comment by Naheel-Azawy** [comment]
@gokcehan Hello again! Sorry for disappearing for long... I Think, unfortunately, I can't work on it soon as I'm getting extremely busy lately... I hope to see some progress on this soon! Thanks!

**Comment by LukeSmithxyz** [comment]
I don't have anything contentful to add to this, but I've been really hoping for this addition for a while. I hope you can find it in your time to complete it sometime soon. I'm dying to uninstall ranger for good and this is the last holdback.

**Comment by Naheel-Azawy** [comment]
@LukeSmithxyz I really wish I have time to work on this... I'm doing my master's, jumping between jobs, and busy with tons of different things... Maybe if @gokcehan can give this issue a higher priority? Maybe @LukeSmithxyz can make a little video about this maybe someone is interested and have time to do it?
Anyway, for me I have left ranger long time ago. Even though I really want this, it is still possible to survive with ascii previews and also with the help of some sxiv magic. Check my [sxiv-key-handler.sh](https://github.com/Naheel-Azawy/naheel-dotfiles/blob/master/configs/sxiv-key-handler.sh) and [sxivv](https://github.com/Naheel-Azawy/naheel-dotfiles/blob/master/scripts/sxivv) for some integration magic between lf and sxiv. I guess you might need my sxiv fork too...

**Comment by cirrusUK** [comment]
setting this as preview script seems to work for images  [sxiv-rifle.sh](https://paste.rs/Wfv.sh)

set previewer /usr/local/bin/sxiv-rifle

File: app.go [hunk 1]
```diff
@@ -32,7 +32,9 @@ type app struct {
 
 func newApp() *app {
 	ui := newUI()
-	nav := newNav(ui.wins[0].h)
+	// TODO: is this really the right way?
+	lastWin := ui.wins[len(ui.wins)-1]
```
Comment (dumblob) [inline app.go:36]: Hm, can't this be influenced by `set preview`/`set nopreview` and/or other (in)direct options (`set-ratios...`)?

**Comment by gokcehan** [comment]
This PR is superseeded by #304 and now is maintained in a separate fork [here](https://gitlab.com/Provessor/lfp) by @Provessor until there is a stable implementation. Closing this PR. Thank you @Naheel-Azawy for starting the work here.

**Pull request closed without merging**